File: leetcode/32. Longest Valid Parentheses/solution.py
class Solution:
    def longestValidParentheses(self, s: 'str') -> 'int':
        stack = [-1]
        r = 0
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
            else:
                stack.pop()
                if stack:
                    r = max(r, i - stack[-1])
                else:
                    stack.append(i)
        return r

File: leetcode/32. Longest Valid Parentheses/solution.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
    var stack = [-1,];
    var r = 0;
    for (var i = 0; i < s.length; i++) {
        if (s.charAt(i) === '(') {
            stack.push(i);
        } else {
            stack.pop();
            if (stack.length === 0) {
                stack.push(i);
            } else {
                r = Math.max(r, i - stack[stack.length - 1]);
            }
        }
    }
    return r;
};
